### PR TITLE
New playground

### PR DESCRIPTION
**Pull request by duailibe**
Context: #3352

This is still WIP and just a POC of a Preact app with no build steps (I opened https://github.com/facebook/Docusaurus/issues/292 to discuss how to reuse Docusaurus infrastructure - then we could use classes and stuff). This is very early but it was usable (commit is a few weeks old, before the big refactor).

I'm opening before finishing to gather feedback: should we keep the "no build steps" for this? It's nice having only one command and not as nice having to use two separate workflows when working on the site (one for docusarus part, other for the playground only) but it's doable.. Ideally I'd want a way to integrate with docusaurus infrastructure (transpilation, dev server, React itself).. anyway, tell me what you think!

### CODE REVIEW

**Comment by duailibe** [comment]
Something's wrong in the preview (as I said, commit is pretty old) will investigate

**Comment by azz** [comment]
Would be really cool if Docusaurus worked like a GatsbyJS plugin. Keep the SSR, single build step, but ability to hydrate the DOM for the playground.

**Comment by duailibe** [comment]
@azz Yeah that's the gist of what I suggested in https://github.com/facebook/Docusaurus/issues/292

**Comment by lipis** [comment]
Can you deploy this somewhere, so people could test it easier?

**Comment by j-f1** [comment]
@lipis https://deploy-preview-3573--prettier.netlify.com/playground/

**Comment by lipis** [comment]
screen shot 2017-12-29 at 17 23 57

**Comment by j-f1** [comment]
> Something's wrong in the preview (as I said, commit is pretty old) will investigate

From @duailibe [above](https://github.com/prettier/prettier/pull/3573#issuecomment-353965310).

**Comment by ikatyang** [comment]
That's #3581, should be fine now.

**Comment by duailibe** [comment]
I merged the master branch, let's see if that fixes it

**Comment by j-f1** [comment]
[off-topic]: Could whoever has access to Netlify [make the deploy logs publicly visible](https://www.netlify.com/blog/2017/10/31/introducing-public-deploy-logs-for-open-source-sites/) so contributors can see what wen wrong if the build fails?

**Comment by duailibe** [comment]
Sorry guys, I forgot this is based on #3433 and I removed the reference to it before creating the PR

**Comment by duailibe** [comment]
Yay it's finally working :)

**Comment by j-f1** [comment]
👍, but 3️⃣❓:

1. What happened to the “Debug” section?
2. Would it be possible to add a “Reset to default” button that resets all of the buttons to their defaults?
3. Would it be possible to make the example code a placeholder so you don’t have to delete it before putting your own code in? (or would that be a bad idea?)

**Comment by duailibe** [comment]
1) Not implemented yet
2) Yeah sure.. as a P/React component it's easy to do that =)
3) I don't know.. we can choose to:
   - always use a placeholder
   - save last code used (+ options)
   - always empty

Those are great but are unrelated to this PR (2 & 3).. I mostly wanted to gather feedback if we should continue with no build steps (just serving from `static/`) or if we should put a separate build step for the playground

I think we can add separate issues for those subjects :)

**Comment by duailibe** [comment]
@azz @ikatyang @vjeux do you guys think I should go forward to get this landed as right now, wait for docusaurs to support this or implement a separate build step? Or give up on this at all

**Comment by vjeux** [comment]
I don't personally care either way as long as we have a working playground at the end :)

**Comment by azz** [comment]
I just reviewed the code, it does look a lot better than what we have now. Do you want to bring it back up to date and land it as-is?

_Possibly related_, I was just reading the code of https://github.com/sandhose/prettier-animated-logo which is also written with Preact, but is for the main page. Would be nice to get full `p?react` support with Docusaurus to make integrating that easier. Currently it is rendered to static markup build-time.

**Comment by duailibe** [comment]
Cool!! Thanks for the review! I'll work on it today on later this week :)

I guess when we get full support with Docusaurus, it'll just be a codemod away to use a nicer syntax.

**Comment by duailibe** [comment]
I think this is ready for review now.. it has (almost) feature parity with the current playground, with the exceptions:

- format by range
- placeholders

Main differences with the current playground:

- editor state is now saved in the local storage... for now only: sidebar is open/closed and debug panels (they are not in the URL anymore)
- added a `Reset to defaults` button in the sidebar

File: .eslintrc.yml [hunk 1]
```diff
@@ -24,9 +24,18 @@ rules:
...
   prefer-const: error
   react/no-deprecated: off
+  react/prop-types: off
   strict: error
   symbol-description: error
   yoda:
   - error
   - never
   - exceptRange: true
+overrides:
```
Comment (j-f1) [inline .eslintrc.yml]: You could put this in `website/playground/.eslintrc.yml` instead to avoid having to use the overrides.

**Comment by lipis** [comment]
Placeholders were nice to have (or any other way to have have the sample code).. will they land in another PR if this is about to get merged or we dropping them?

**Comment by duailibe** [comment]
I'm still gonna add them in this PR.. only the range stuff I'm gonna do in another PR

**Comment by duailibe** [comment]
I ended up doing both the examples and the range feature in this PR =)

Gonna merge when lint/tests pass and I do some smoke testing with the deploy preview

**Comment by duailibe** [comment]
I hate when I press merge instead of the squash button 😔

**Comment by j-f1** [comment]
@duailibe Should we disable the other merge types (admins can do it in settings), or are there times we use them  in this repo?

**Comment by duailibe** [comment]
@j-f1 I think we can (TBH I don't think there's much use in doing it other way anyway) cc @vjeux

**Comment by vjeux** [comment]
Sgtm

**Comment by lipis** [comment]
I disabled the other two options (if for whatever reason, we need any other option, we can easily enable it before doing it)

**Pull request merged**